compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -1098,66 +1098,135 @@ private BType inferTableMemberType(List<BType> memTypes, BType expType) {
-            allFieldSet.addAll(((BRecordType) memType).fields.values());
-        }
+        List<BType> restFieldTypes = new ArrayList<>();
 
-        Set<BField> commonFieldSet = new LinkedHashSet<>(allFieldSet);
-        for (BType memType : memTypes) {
-            commonFieldSet.retainAll(((BRecordType) memType).fields.values());
-        }
 
-        List<String> requiredFieldNames = new ArrayList<>();
         if (keySpecifier != null) {
             for (IdentifierNode identifierNode : keySpecifier.fieldNameIdentifierList) {
-                requiredFieldNames.add(((BLangIdentifier) identifierNode).value);
                 keySpecifierFieldNames.add(((BLangIdentifier) identifierNode).value);
             }
         }
 
-        List<String> fieldNames = new ArrayList<>();
-        for (BField field : allFieldSet) {
-            String fieldName = field.name.value;
+        LinkedHashMap<String, List<BField>> fieldNameToFields = new LinkedHashMap<>();
+        for (BType memType : memTypes) {
+            BRecordType member = (BRecordType) memType;
+            for (BField field : member.fields.values()) {
+                String key = field.name.value;
+                if (fieldNameToFields.containsKey(key)) {
+                    fieldNameToFields.get(key).add(field);
+                } else {
+                    fieldNameToFields.put(key, new ArrayList<>() {{
+                        add(field);
+                    }});
+                }
+            }
 
-            if (fieldNames.contains(fieldName)) {
-                dlog.error(tableConstructorExpr.pos,
-                        DiagnosticErrorCode.CANNOT_INFER_MEMBER_TYPE_FOR_TABLE_DUE_AMBIGUITY,
-                        fieldName);
-                return symTable.semanticError;
+            if (!member.sealed) {
+                BType restFieldType = member.restFieldType;
+                if (isUniqueType(restFieldTypes, restFieldType)) {
+                    restFieldTypes.add(restFieldType);
+                }
             }
-            fieldNames.add(fieldName);
+        }
 
-            boolean isOptional = true;
-            for (BField commonField : commonFieldSet) {
-                if (commonField.name.value.equals(fieldName)) {
-                    isOptional = false;
-                    requiredFieldNames.add(commonField.name.value);
+        LinkedHashSet<BField> inferredFields = new LinkedHashSet<>();
+        int memTypesSize = memTypes.size();
+
+        fieldNameToFields.forEach((fieldName, fields) -> {
+            List<BType> uniqueTypes = getUniqueTypesFromSameNameFieldList(fields);
+
+            for (BType memType : memTypes) {
+                BRecordType bMemType = (BRecordType) memType;
+                if (bMemType.sealed) {
+                    continue;
+                }
+
+                if (bMemType.fields.containsKey(fieldName)) {
+                    continue;
+                }
+
+                BType restFieldType = bMemType.restFieldType;
```

Comment:
Wondering if we have to explicitly handle never-typed individual fields and rest fields.

```ballerina
type Foo record {|
    int i;
    never j?;
    never k?;
    never...;
|};

function fn(Foo f) {
    var t = table [
        {...f},
        {i: 1, j: 2, l: ""}
    ];
    int _ = t;
}
```

results in

```
incompatible types: expected 'int', found 'table<record {| int i; (never|int) j?; never k?; (string|never) l?; never...; |}>'
```

`T|never` is `T`, so wondering if we should be saying `never|int`.

Reply:
Then we seem to consider scenarios like `anydata`, `any` also at both places where union type is created.
```
public function main() {
    anydata f = 1;

    var t = table [
        {a: f},
        {a: 1}
    ];
    int _ = t;
}
```
`incompatible types: expected 'int', found 'table<record {| (anydata|int) a; |}>`

The problem is only when there is an error thrown at the user. Otherwise, with correct assignments, it works right?

Wondering whether this is affecting only changes introduced in this PR. What about all other places we create union type from a list. I kind of feel like there should be a common API to create a union while doing these validations.

Will update the PR.

Reply:
Both cases are fixed.